### PR TITLE
Add patience argument to Trainer

### LINKED ISSUE #4894: 🚀 Add early stopping to the trainer
# 🚀 Feature request

The trainer ([pt](https://github.com/huggingface/transformers/blob/master/src/transformers/trainer.py), [tf](https://github.com/huggingface/transformers/blob/master/src/transformers/trainer_tf.py)) is an easy access point for users who rather not spend too much time building their own trainer class but prefer an out-of-the-box solution. Even though `transformers` was never meant to be a fully fletched training library, it might please users to add an additional feature: early stopping. 

## Motivation

Early stopping ensures that the trainer does not needlessly keep training when the loss does not improve. This saves time, money, and let's not forget the trees. 😉 Performance-wise this should not lead to different results.

## Your contribution

At the moment I cannot work on this, but here are my thoughts:

- a training argument should be added ([pt](https://github.com/huggingface/transformers/blob/master/src/transformers/training_args.py), [tf](https://github.com/huggingface/transformers/blob/master/src/transformers/trainer_tf.py)). This would only work when `evaluate_during_training` is enabled.
- for PyTorch: at every evaluation step, an early stopper (can be a separate class even) checks if the loss has improved in the last n steps. Potentially with a minimal threshold that the loss should have improved. If not, the trainer should stop
- for Tensorflow: I don't have experience with TF myself, but I assume one could use [`tf.keras.callbacks.EarlyStopping`](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/EarlyStopping).

### PR DESCRIPTION
**Pull request by thesamuel**
This closes #4894.

# Summary
Often, we want to stop training if loss does not improve for a number of epochs. This PR adds a "patience" argument, which is a limit on the number of times we can get a non-improving eval loss before stopping training early.

It is implemented by other NLP frameworks, such as AllenNLP (see [trainer.py](https://github.com/allenai/allennlp/blob/master/allennlp/training/trainer.py#L95) and [metric_tracker.py](https://github.com/allenai/allennlp/blob/1a8a12cd1b065d74fec3d2e80105a684736ff709/allennlp/training/metric_tracker.py#L6)).

# Motivation
This feature allows faster fine-tuning by breaking the training loop early and avoids users the toil of checking metrics on Tensorboard.

# Caveats
Often, models are evaluated once per epoch, but run_lm_finetuning.py has an option to evaluate after a set number of model update steps (dictated by `--logging_steps` if `--evaluate_during_training` is true). Because of this, I've elected to tie patience to the number of evaluations without improvement in loss.

### CODE REVIEW

**Comment by thesamuel** [comment]
This supercedes https://github.com/huggingface/transformers/pull/2840, where I added patience to the outdated `run_language_modeling.py` script.

**Comment by BramVanroy** [comment]
Looking good! Can you add a reference to your original post that this closes https://github.com/huggingface/transformers/issues/4894? Thanks

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -307,6 +309,8 @@ def train(self, model_path: Optional[str] = None):
 
         tr_loss = 0.0
         logging_loss = 0.0
+        best_eval_loss = None
+        evals_without_improvement = 0
```
Comment (julien-c) [inline src/transformers/trainer.py:313]: nit: prefix those with `patience_` as they're specific to this

```suggestion
        patience_best_eval_loss = None
        patience_evals_without_improvement = 0
        patience_should_stop = False
```

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -347,6 +351,18 @@ def train(self, model_path: Optional[str] = None):
...
+                                    eval_loss = results["loss"]
+                                    if not best_eval_loss or eval_loss < best_eval_loss:
+                                        evals_without_improvement = 0
+                                        best_eval_loss = eval_loss
+                                    else:
+                                        evals_without_improvement += 1
+                                        if evals_without_improvement >= self.args.patience:
+                                            logger.info(
+                                                f"Patience threshold ({self.args.patience}) exceeded, stopping training"
+                                            )
```
Comment (julien-c) [inline src/transformers/trainer.py:365]:
```suggestion
                                            patience_should_stop = True
                                            logger.info(
                                                f"Patience threshold ({self.args.patience}) exceeded, stopping training"
                                            )
```

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -377,10 +393,12 @@ def train(self, model_path: Optional[str] = None):
                             torch.save(scheduler.state_dict(), os.path.join(output_dir, "scheduler.pt"))
                             logger.info("Saving optimizer and scheduler states to %s", output_dir)
 
-                if self.args.max_steps > 0 and global_step > self.args.max_steps:
+                if ((self.args.max_steps > 0 and global_step > self.args.max_steps) or
+                        (self.args.patience > 0 and evals_without_improvement >= self.args.patience)):
```
Comment (julien-c) [inline src/transformers/trainer.py:397]:
```suggestion
                if ((self.args.max_steps > 0 and global_step > self.args.max_steps) or
                        patience_should_stop):
```

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -377,10 +393,12 @@ def train(self, model_path: Optional[str] = None):
...
                     epoch_iterator.close()
                     break
-            if self.args.max_steps > 0 and global_step > self.args.max_steps:
+            if ((self.args.max_steps > 0 and global_step > self.args.max_steps) or
+                    (self.args.patience > 0 and evals_without_improvement >= self.args.patience)):
```
Comment (julien-c) [inline src/transformers/trainer.py:401]: same here

**Review by julien-c — CHANGES_REQUESTED** [review]
Looks good! Small suggestions there

**Comment by kevin-yauris** [comment]
Hello, when this feature will be merged? I would like to use it. Thank you.

**Comment by BramVanroy** [comment]
> Hello, when this feature will be merged? I would like to use it. Thank you.

There are some changes requested that @thesamuel should fix before this can be merged.

**Comment by misrasaurabh1** [comment]
Bump. Early stopping is critical for an automated Trainer that reliably gives us the best model. Current way of figuring out the training stopping point seems to be specifying a static train_epochs but the training duration a model can take depends on way too many factors like learning rate, data complexity, model, model size, optimizer and so on that it is unreasonable to ask the user to specify the epochs in advance.
I believe the current assumption is that people train with very small learning rates so that the loss always seems to keep decreasing very slowly but according to my experience (and on my data) it is a sub-optimal schedule which takes too much time. I see that training with higher learning rates with larger batch sizes and stopping at the early stopping point results in an equally good if not better models. Although this requires use of early stopping.

**Comment by PhilipMay** [comment]
I would like to use this early stopping on downstream training.
The current implementation only stops training by monitoring loss. IMO it should also be possible to monitor other metrics like F1 and ROC-AUC. 

I also would like to add a feature that stores the model each time when the monitored metric improves and then optionaly loads the model it after training. Then later evaluation can be done on this "best" model.

@thesamuel @julien-c what do you think?

**Comment by sgugger** [comment]
I plan to work on this once I'm finished with the Funnel Transformer model @PhilipMay (so end of this week, beginning of the next).

**Comment by PhilipMay** [comment]
> 
> 
> I plan to work on this once I'm finished with the Funnel Transformer model @PhilipMay (so end of this week, beginning of the next).

@sgugger That would be awsome. Maybe you want to get some inspiration from the FARM training loop which is pretty nice IMO:

https://github.com/PhilipMay/FARM/blob/nested_xval/farm/train.py#L262-370

**Comment by PhilipMay** [comment]
I just found this PR that was already merged: #7431
I think it solved this...

**Comment by sgugger** [comment]
Not quite, but it makes implementing it easier.

**Comment by PhilipMay** [comment]
> Not quite, but it makes implementing it easier.

Yes - you are right. The patience part is still missing.

**Comment by BramVanroy** [comment]
@sgugger Should we keep this open? You wrote in this thread you will work on this if you find the time, but I am not sure if you plan to use another PR for that.

**Comment by sgugger** [comment]
There has been a PR merged adding the `EarlyStoppingCallback` (#8581) so I think this can be closed now.

**Comment by thesamuel** [comment]
Thanks @cbrochtrup @sgugger! Sorry I didn't get around to this...

**Comment by cbrochtrup** [comment]
You're welcome, happy to help!
